src/main/kotlin/org/rust/ide/ssr/RsMatchingVisitor.kt

```diff
@@ -0,0 +1,258 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.ssr
+
+import com.intellij.psi.PsiElement
+import com.intellij.structuralsearch.impl.matcher.GlobalMatchingVisitor
+import com.intellij.structuralsearch.impl.matcher.handlers.SubstitutionHandler
+import org.rust.ide.utils.import.ImportCandidatesCollector
+import org.rust.ide.utils.import.ImportContext
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.psi.ext.stringValue
+
+class RsMatchingVisitor(private val myMatchingVisitor: GlobalMatchingVisitor) : RsVisitor() {
+    private fun getHandler(element: PsiElement) = myMatchingVisitor.matchContext.pattern.getHandler(element)
+
+    private fun matchTextOrVariable(templateElement: PsiElement?, treeElement: PsiElement?): Boolean {
+        if (templateElement == null) return true
+        if (treeElement == null) return false
+        return when (val handler = getHandler(templateElement)) {
+            is SubstitutionHandler -> handler.validate(treeElement, myMatchingVisitor.matchContext)
+            else -> myMatchingVisitor.matchText(templateElement, treeElement)
+        }
+    }
+
+    private inline fun <reified T> getElement(): T? = when (val element = myMatchingVisitor.element) {
+        is T -> element
+        else -> {
+            myMatchingVisitor.result = false
+            null
+        }
+    }
+
+    private fun match(e1: PsiElement?, e2: PsiElement?) = myMatchingVisitor.match(e1, e2)
+
+    override fun visitStructItem(o: RsStructItem) {
+        val struct = getElement<RsStructItem>() ?: return
+        myMatchingVisitor.result =
+            matchOuterAttrList(o.outerAttrList, struct.outerAttrList) &&
+                match(o.vis, struct.vis) &&
+                matchIdentifier(o.identifier, struct.identifier) &&
+                match(o.typeParameterList, struct.typeParameterList) &&
+                match(o.whereClause, struct.whereClause) &&
+                match(o.blockFields, struct.blockFields) &&
+                match(o.tupleFields, struct.tupleFields)
+    }
+
+    override fun visitTypeParameterList(o: RsTypeParameterList) {
+        val parameters = getElement<RsTypeParameterList>() ?: return
+        myMatchingVisitor.result =
+            myMatchingVisitor.matchSequentially(o.typeParameterList, parameters.typeParameterList) &&
+                myMatchingVisitor.matchSequentially(o.lifetimeParameterList, parameters.lifetimeParameterList) &&
+                myMatchingVisitor.matchSequentially(o.constParameterList, parameters.constParameterList)
+    }
+
+    override fun visitWhereClause(o: RsWhereClause) {
```

Comment:
I think it is ok not to match equivalent constructs and wait for a feature request.
In a custom inspection we can add another template to match both constructs.

Reply:
I also think we should not merge such cases (at least by default) because SSR is about syntax, not about semantics.
For example, you may want to have a custom inspection to annotate all usages of `struct Foo1<T: Sized>(T);` style and a quick-fix to convert them to `struct Foo2<T>(T) where T: Sized;` style